File: port/cpp/probe/belief_probe.py
"""Runs the same cases through the Python belief engine, and compares.

The C++ side prints its belief store with every double as a raw bit pattern.
This runs `belief_cases.json` through `unscripted.belief` -- the real module, not a
re-implementation -- serialises it the same way, and diffs the two.

Naming the differing field matters more than counting them. A port's first run
disagrees in a dozen places, and a wall of JSON says nothing about which one to
fix; `one_origin_repeated.beliefs.was_at(...).logit` says exactly where to look
and, usually, why.

    python3 port/cpp/probe/belief_probe.py
"""
from __future__ import annotations

import json
import os
import struct
import subprocess
import sys
import tempfile

HERE = os.path.dirname(os.path.abspath(__file__))
PORT_ROOT = os.path.dirname(HERE)
REPO_ROOT = os.path.dirname(os.path.dirname(PORT_ROOT))
CASES = os.path.join(HERE, "belief_cases.json")

sys.path.insert(0, REPO_ROOT)

from unscripted.belief import BeliefEngine                     # noqa: E402
from unscripted.ontology import Proposition                    # noqa: E402


def to_bits(value):
    """Every float in a blob rewritten as its raw 64 bits.

    Floats only. An int stays an int, because `1` and `1.0` are a real
    difference here: one of them means a port produced an integer where the
    reference produced a float, and that shows up in the exported state.
    """
    if isinstance(value, float):
        return "b:%016x" % struct.unpack("<Q", struct.pack("<d", value))[0]
    if isinstance(value, (list, tuple)):
        # TUPLES TOO. A reason detail carries `[(action, probability), ...]` as
        # tuples, and a tuple that fell through here kept its floats as decimal
        # numbers while the C++ side had already turned its arrays into bit
        # patterns -- so the two could never match and the mismatch read as a
        # value difference rather than as a type one.
        return [to_bits(item) for item in value]
    if isinstance(value, dict):
        return {key: to_bits(item) for key, item in value.items()}
    return value


def python_side(document: dict) -> dict:
    engine = BeliefEngine()
    report = {}
    for scenario in document["cases"]:
        beliefs, traces = {}, []
        for update in scenario["updates"]:
            repeat = update.get("repeat", 1)
            for step in range(repeat):
                claim = Proposition(update["predicate"], dict(update["slots"]),
                                    "-" if update.get("store_negated") else "+")
                origin = update.get("origin")
                if origin == "vary":
                    origin = step
                claim_id = update["claim_id"]
                if repeat > 1:
                    claim_id = "%s:%d" % (claim_id, step)
                _belief, reasons = engine.update(
                    beliefs, claim, asserter_polarity=update["polarity"],
                    trust=update["trust"], competence=update["competence"],
                    skepticism=update["skepticism"], claim_id=claim_id,
                    origin_event=origin, world_time=update["world_time"],
                    speaker=update.get("speaker"))
                if step + 1 == repeat:
                    traces.append([{"code": code, "amount": amount, "detail": detail}
                                   for code, amount, detail in reasons])
        report[scenario["id"]] = to_bits({
            "beliefs": {key: belief.as_dict() for key, belief in beliefs.items()},
            "reasons": traces,
        })
    return report


def cpp_side() -> dict:
    compiler = os.environ.get("CXX", "g++")
    with tempfile.TemporaryDirectory() as work:
        binary = os.path.join(work, "belief_probe")
        subprocess.run(
            [compiler, "-O2", "-std=c++17", "-I", os.path.join(PORT_ROOT, "include"),
             os.path.join(HERE, "belief_probe.cpp"), "-o", binary],
            check=True)
        out = subprocess.run([binary, CASES], check=True, capture_output=True, text=True)
    return json.loads(out.stdout)


def differences(expected, actual, path="") -> list:
    """Field paths where two blobs disagree, deepest name first."""
    if type(expected) is not type(actual):
        return [(path, _short(expected), _short(actual))]
    if isinstance(expected, dict):
        out = []
        for key in list(expected) + [k for k in actual if k not in expected]:
            where = f"{path}.{key}" if path else key
            if key not in expected:
                out.append((where, "<absent>", _short(actual[key])))
            elif key not in actual:
                out.append((where, _short(expected[key]), "<absent>"))
            else:
                out += differences(expected[key], actual[key], where)
        # Order is part of the answer, not a detail: `primary_origin` is the
        # first key of `origin_counts`, and `core_key` walks slots in order.
        if list(expected) != list(actual) and set(expected) == set(actual):
            out.append((path or "<root>", "key order " + ", ".join(list(expected)[:4]),
                        "key order " + ", ".join(list(actual)[:4])))
        return out
    if isinstance(expected, list):
        if len(expected) != len(actual):
            return [(path + "[]", "%d items" % len(expected), "%d items" % len(actual))]
        out = []
        for index, (one, two) in enumerate(zip(expected, actual)):
            out += differences(one, two, f"{path}[{index}]")
        return out
    if expected != actual:
        return [(path, _short(expected), _short(actual))]
    return []


def _short(value) -> str:
    text = value if isinstance(value, str) else json.dumps(value, sort_keys=False)
    return text if len(text) <= 72 else text[:69] + "..."


def main() -> int:
    with open(CASES, encoding="utf-8") as handle:
        document = json.load(handle)

    expected = python_side(document)
    try:
        actual = cpp_side()
    except FileNotFoundError:
        print("no C++ compiler; nothing measured")
        return 0
    except subprocess.CalledProcessError as exc:
        print("the probe would not build or run: %s" % exc)
        return 1

    total = clean = 0
    for scenario in document["cases"]:
        name = scenario["id"]
        total += 1
        rows = differences(expected.get(name), actual.get(name))
        if not rows:
            clean += 1
            print("  ok    %s" % name)
            continue
        print("  DIFF  %s   %d field(s)" % (name, len(rows)))
        for where, exp, act in rows[:6]:
            print("          %s" % where)
            print("            python %s" % exp)
            print("            c++    %s" % act)
        if len(rows) > 6:
            print("          ... and %d more" % (len(rows) - 6))

    print("\n%d of %d cases reproduce the Python belief engine exactly." % (clean, total))
    if clean != total:
        return 1
    print("Beliefs, provenance, origin counts, contradictions and reason traces "
          "all agree,\nincluding the order of every map.")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: port/cpp/probe/belief_probe.cpp
// Runs `belief_cases.json` through the ported belief engine and prints the
// resulting belief store as JSON, with every double written as its raw 64 bits.
//
// Bits rather than decimals, for the same reason `bit_probe` does it: two
// numbers that print identically to fifteen places can still be different
// doubles, and the difference between them is the difference between two worlds
// a thousand ticks later.
#include <cstdint>
#include <cstdio>
#include <cstring>
#include <fstream>
#include <sstream>
#include <string>

#include "usc/belief.hpp"
#include "usc/json.hpp"

namespace {

std::string read_file(const char* path) {
    std::ifstream file(path);
    std::ostringstream buffer;
    buffer << file.rdbuf();
    return buffer.str();
}

/// A double as its raw bits, tagged so it cannot be mistaken for a number that
/// happened to be written the same way.
usc::Json bits_of(double value) {
    std::uint64_t raw;
    std::memcpy(&raw, &value, sizeof raw);
    char buffer[32];
    std::snprintf(buffer, sizeof buffer, "b:%016llx",
                  static_cast<unsigned long long>(raw));
    return usc::Json(std::string(buffer));
}

/// Rewrite every number in a blob as its bit pattern, recursively.
usc::Json to_bits(const usc::Json& value) {
    switch (value.kind()) {
        case usc::Json::Kind::Real:
            return bits_of(value.as_double());
        case usc::Json::Kind::Array: {
            usc::Json out = usc::Json::array();
            for (const auto& item : value.items()) out.push(to_bits(item));
            return out;
        }
        case usc::Json::Kind::Object: {
            usc::Json out = usc::Json::object();
            for (const auto& entry : value.fields())
                out.fields()[entry.first] = to_bits(entry.second);
            return out;
        }
        default:
            return value;
    }
}

usc::Proposition proposition_from(const usc::Json& update) {
    usc::Proposition prop;
    prop.predicate = update.get("predicate").as_string();
    const usc::Json* slots = update.find("slots");
    if (slots) for (const auto& entry : slots->fields()) prop.slots[entry.first] = entry.second;
    prop.polarity = "+";
    return prop;
}

}  // namespace

int main(int argc, char** argv) {
    const char* path = argc > 1 ? argv[1] : "belief_cases.json";
    usc::Json document = usc::Json::parse(read_file(path));

    usc::Json report = usc::Json::object();
    usc::BeliefEngine engine;

    for (const usc::Json& scenario : document.at("cases").items()) {
        usc::OrderedMap<usc::Belief> beliefs;
        usc::Json traces = usc::Json::array();

        for (const usc::Json& update : scenario.at("updates").items()) {
            long long repeat = 1;
            if (const usc::Json* found = update.find("repeat")) repeat = found->as_int(1);

            for (long long step = 0; step < repeat; ++step) {
                usc::Proposition claim = proposition_from(update);
                // `store_negated` files the claim under its own core key while
                // asserting the opposite -- how a denial of a distinct
                // proposition reaches the contradiction scan.
                if (update.get("store_negated").as_bool()) claim.polarity = "-";

                usc::Json origin = update.get("origin");
                // "vary" means a fresh origin per repetition: the case is about
                // the provenance trim, not about the correlation discount.
                if (origin.kind() == usc::Json::Kind::String && origin.as_string() == "vary")
                    origin = usc::Json(static_cast<long long>(step));

                usc::Json claim_id = update.get("claim_id");
                if (repeat > 1)
                    claim_id = usc::Json(claim_id.as_string() + ":" + std::to_string(step));

                std::vector<usc::Reason> reasons = engine.update(
                    beliefs, claim, update.get("polarity").as_string(),
                    update.get("trust").as_double(), update.get("competence").as_double(),
                    update.get("skepticism").as_double(), claim_id, origin,
                    update.get("world_time"), update.get("speaker"));

                // Only the last trace of a repeated update is kept, so a case
                // with 400 repetitions does not bury the comparison.
                if (step + 1 == repeat) {
                    usc::Json trace = usc::Json::array();
                    for (const usc::Reason& reason : reasons) {
                        usc::Json row = usc::Json::object();
                        row["code"] = reason.code;
                        row["amount"] = reason.amount;
                        row["detail"] = reason.detail;
                        trace.push(row);
                    }
                    traces.push(trace);
                }
            }
        }

        usc::Json store = usc::Json::object();
        for (const auto& entry : beliefs)
            store.fields()[entry.first] = entry.second.as_json();

        usc::Json result = usc::Json::object();
        result["beliefs"] = store;
        result["reasons"] = traces;
        report.fields()[scenario.get("id").as_string()] = to_bits(result);
    }

    std::printf("%s\n", report.dump(2).c_str());
    return 0;
}
